lib/csv.rb

```diff
@@ -769,33 +769,60 @@ def instance(data = $stdout, **options)
       end
     end
 
-    #
     # :call-seq:
-    #   filter( **options ) { |row| ... }
-    #   filter( input, **options ) { |row| ... }
-    #   filter( input, output, **options ) { |row| ... }
-    #
-    # This method is a convenience for building Unix-like filters for CSV data.
-    # Each row is yielded to the provided block which can alter it as needed.
-    # After the block returns, the row is appended to +output+ altered or not.
-    #
-    # The +input+ and +output+ arguments can be anything CSV::new() accepts
-    # (generally String or IO objects). If not given, they default to
-    # <tt>ARGF</tt> and <tt>$stdout</tt>.
-    #
-    # The +options+ parameter is also filtered down to CSV::new() after some
-    # clever key parsing. Any key beginning with <tt>:in_</tt> or
-    # <tt>:input_</tt> will have that leading identifier stripped and will only
-    # be used in the +options+ Hash for the +input+ object. Keys starting with
-    # <tt>:out_</tt> or <tt>:output_</tt> affect only +output+. All other keys
-    # are assigned to both objects.
-    #
-    # See {Options for Parsing}[#class-CSV-label-Options+for+Parsing]
-    # and {Options for Generating}[#class-CSV-label-Options+for+Generating].
-    #
-    # The <tt>:output_row_sep</tt> +option+ defaults to
-    # <tt>$INPUT_RECORD_SEPARATOR</tt> (<tt>$/</tt>).
+    #   filter(**options) {|row| ... } -> integer
```

Comment:
Can we say that "the return value is nothing"?
The current `#filter` implementation returns an integer but it just so happens that `input.each` returns integer.
We don't want users to expect `#filter` returns an integer.

Reply:
I agree that this is weird.  We see this in a number of places.  Can we return something useful when possible, or nil otherwise?

Reply:
I think that it's enough that we say it in document.
Should we return `nil` explicitly? If we should return something, `#filter` should return `nil`.

Reply:
::filter could return the output CSV object or both CSV objects.
::foreach could return the CSV object.
::parse could return the CSV object.

What the user would do with these is none of our business?

Reply:
Or just nil everywhere we're now returning an integer.

Reply:
I have changed the returns from integer to nil for the three methods above.

Reply:
This is no longer a doc-only PR.

Reply:
@kou, can you take a look?

Reply:
I don't want to state that returning `nil` or something is the our spec for now.

In the current our knowledge, we don't have enough reason to state that `#filter` should return `nil`, the output CSV object or both CSV objects. We may find enough reason to state that `#filter` should return XXX in the future. If we state that `#filter` should return ZZZ now, we need to break backward compatibility.

So I think that we should state that `#filter` returns nothing or `#filter`'s return value is undefined for now.
What do you think about this?

Reply:
Works for me.  I'll remove the nil returns and delete the nil returns from the call-seqs.
Some of them say "thingy or nil";  I'll just say "thingy".  OK/

Reply:
Done.